components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+        while let Some(result) = results.next() {
+            let row = result?;
+            let guid = row.get_checked::<_, SyncGuid>("guid")?;
+            let kind = SyncedBookmarkKind::from_u8(row.get_checked("kind")?)?;
+            let mut item = Item::new(guid.into(), kind.into());
+            let age = ServerTimestamp(row.get_checked::<_, f64>("serverModified").unwrap_or(0f64))
+                .duration_since(self.remote_time)
+                .unwrap_or_default();
+            item.age = age.as_secs() as i64 * 1000 + i64::from(age.subsec_millis());
+            item.needs_merge = row.get_checked("needsMerge")?;
+            item.validity = SyncedBookmarkValidity::from_u8(row.get_checked("validity")?)?.into();
+
+            let p = builder.item(item)?;
+            if let Some(parent_guid) = row.get_checked::<_, Option<SyncGuid>>("parentGuid")? {
+                p.by_parent_guid(parent_guid.into())?;
+            }
+        }
+
+        let sql = format!(
+            "
+            SELECT guid, parentGuid FROM moz_bookmarks_synced_structure
+            WHERE guid <> '{root_guid}'
+            ORDER BY parentGuid, position",
+            root_guid = BookmarkRootGuid::Root.as_guid().as_ref()
+        );
+        let mut stmt = self.db.prepare(&sql)?;
+        let mut results = stmt.query(NO_PARAMS)?;
+        while let Some(result) = results.next() {
+            let row = result?;
+            let guid = row.get_checked::<_, SyncGuid>("guid")?;
+            let parent_guid = row.get_checked::<_, SyncGuid>("parentGuid")?;
+            builder
+                .parent_for(&guid.into())
+                .by_children(&parent_guid.into())?;
+        }
+
+        let mut tree = builder.into_tree()?;
+
+        // Note tombstones for remotely deleted items.
+        let mut stmt = self
+            .db
+            .prepare("SELECT guid FROM moz_bookmarks_synced WHERE isDeleted AND needsMerge")?;
+        let rows = stmt.query_and_then(NO_PARAMS, |row| row.get_checked::<_, SyncGuid>("guid"))?;
+        for row in rows {
+            let guid = row?;
+            tree.note_deleted(guid.into());
+        }
+
+        Ok(tree)
+    }
+
+    /// Fetches content info for all synced items that changed since the last
+    /// sync and don't exist locally.
+    fn fetch_new_remote_contents(&self) -> Result<HashMap<dogear::Guid, Content>> {
+        let mut contents = HashMap::new();
+
+        let sql = format!(
+            r#"
+            SELECT v.guid, v.kind, IFNULL(v.title, "") AS title, h.url,
+                   s.position
+            FROM moz_bookmarks_synced v
+            JOIN moz_bookmarks_synced_structure s ON s.guid = v.guid
+            LEFT JOIN moz_places h ON h.id = v.placeId
+            LEFT JOIN moz_bookmarks b ON b.guid = v.guid
+            WHERE NOT v.isDeleted AND
+                  v.needsMerge AND
+                  b.guid IS NULL AND
+                  s.parentGuid <> '{root_guid}'"#,
```

Comment:
`AND validity <> SyncedBookmarkValidity::Replace`, too; we can't meaningfully dedupe invalid items. (In fact, I think that will cause `fetch_new_remote_contents` to return an error now, since `url` will be null).